File: android/app/src/main/java/ru/cmpas/app/presentation/util/SlotFormat.kt
package ru.cmpas.app.presentation.util

/**
 * Годится ли слот расписания выбранному формату встречи.
 *
 * Правило задано учредителем: «В кабинете» — только очные слоты, «Онлайн» —
 * только онлайновые, а гибридное правило расписания («Онлайн + Кабинет»)
 * показывается и там и там.
 *
 * До этого отбора не было вовсе: экран «Добавить запись» показывал все
 * свободные часы дня независимо от выбранного формата, и специалист,
 * выбравший «В кабинете», видел среди них утренние онлайновые.
 *
 * Словари форматов в системе два: расписание пишет «offline», сессия —
 * «in_person». Сравнивать их напрямую нельзя: значения разные, а смысл
 * один, и молчаливое несовпадение спрятало бы все очные слоты вместо
 * лишних. Поэтому оба приводятся к одному виду здесь.
 */
object SlotFormat {

    const val ONLINE = "online"
    const val OFFLINE = "offline"

    /** Гибрид: правило допускает обе формы встречи. */
    const val BOTH = "both"

    /** «in_person», «IN_PERSON», «offline» — одно и то же. */
    fun normalize(value: String?): String = when (value?.trim()?.lowercase()) {
        "offline", "in_person", "inperson" -> OFFLINE
        BOTH -> BOTH
        else -> ONLINE
    }

    /**
     * @param slotFormat формат слота, как его отдал сервер
     * @param chosenFormat формат, выбранный на экране
     */
    fun matches(slotFormat: String?, chosenFormat: String?): Boolean {
        val slot = slotFormat?.trim()?.lowercase()
        // Гибрид годится всегда — это и есть решение учредителя про
        // «гибридное расписание выдаёт и те и другие».
        if (slot == BOTH) return true
        return normalize(slotFormat) == normalize(chosenFormat)
    }
}

File: src/lib/booking/slot-format.ts
/**
 * Годится ли слот расписания выбранному формату встречи.
 *
 * Правило: «в кабинете» — только очные слоты, «онлайн» — только
 * онлайновые, гибридное правило расписания («Онлайн + Кабинет») годится и
 * туда и туда.
 *
 * Отбора не было ни в приложении, ни в вебе: обе формы записи показывали
 * все свободные часы дня подряд, независимо от выбранного формата.
 * Специалист, выбравший «в кабинете», видел среди них утренние
 * онлайновые.
 *
 * Словари форматов в системе два: расписание пишет «offline», сессия —
 * «in_person». Сравнивать напрямую нельзя — значения разные, смысл один, и
 * молчаливое несовпадение спрятало бы ВСЕ очные слоты вместо лишних, то
 * есть починка дала бы пустой экран вместо неверного.
 */

export const SLOT_ONLINE = 'online';
export const SLOT_OFFLINE = 'offline';
/** Гибрид: правило допускает обе формы встречи. */
export const SLOT_BOTH = 'both';

/** «in_person», «IN_PERSON», «offline» — одно и то же. */
export function normalizeSlotFormat(value: string | null | undefined): string {
    switch ((value || '').trim().toLowerCase()) {
        case 'offline':
        case 'in_person':
        case 'inperson':
            return SLOT_OFFLINE;
        case SLOT_BOTH:
            return SLOT_BOTH;
        default:
            return SLOT_ONLINE;
    }
}

export function slotMatchesFormat(slotFormat: string | null | undefined, chosenFormat: string | null | undefined): boolean {
    // Гибрид годится всегда.
    if ((slotFormat || '').trim().toLowerCase() === SLOT_BOTH) return true;
    return normalizeSlotFormat(slotFormat) === normalizeSlotFormat(chosenFormat);
}
